src/components/FacetController.js

```diff
@@ -23,202 +30,147 @@ function validFrameProps(originalProps, frameType) {
+      if (facetHoverAnnotation.type === "column-hover") {
+        facetHoverAnnotation.facetColumn = facetHoverAnnotation.column.name
+        facetHoverAnnotation.column = undefined
+      } else {
+        facetHoverAnnotation.type = "frame-hover"
+        facetHoverAnnotation.y = undefined
+        facetHoverAnnotation.yBottom = undefined
+        facetHoverAnnotation.yMiddle = undefined
+        facetHoverAnnotation.yTop = undefined
+      }
 
-      yExtentPromises.push(extentPromise)
-      rest.yExtent = { onChange: extentCallback }
+      annotations.push(facetHoverAnnotation)
     }
 
-    return React.cloneElement(child, validFrameProps(rest, frameType))
+    return annotations
   }
 
-  mapChildrenWithHover = (child, childI) => {
-    const annotations = [...this.state.baseChildAnnotations[childI]] || []
+  /**
+   * Map hover annotations and extent to child. Initially the extent is an object with
+   * an onChange handler however once each of those resolve we then create an
+   * extent that matches between all of them. This logic can be found in createExtent and also
+   * extentHandler
+   */
+  mapChildrenWithAppropriateProps = ({
+    child,
+    props,
+    state
+  }: {
+    child: Element<*>,
+    props: Props,
+    state: State
+  }) => {
+    const frameType = child.type.displayName
+    const annotations = this.mapChildrenWithHoverAnnotations({ child, state })
+    const customProps = { ...props, annotations }
 
-    if (this.state.facetHover) {
-      const facetHoverAnnotation = { ...this.state.facetHover }
-      if (this.state.facetHover.type === "column-hover") {
-        delete facetHoverAnnotation.column
-        facetHoverAnnotation.facetColumn = this.state.facetHover.column.name
-      } else {
-        facetHoverAnnotation.type = "frame-hover"
-        delete facetHoverAnnotation.y
-        delete facetHoverAnnotation.yBottom
-        delete facetHoverAnnotation.yMiddle
-        delete facetHoverAnnotation.yTop
-      }
+    // pieceHoverAnnotation could be an object, so we need to be explicit in checking for true
+    if (props.hoverAnnotation === true || props.pieceHoverAnnotation === true) {
+      customProps.customHoverBehavior = d => this.setState({ facetHover: d })
+    }
 
-      annotations.push(facetHoverAnnotation)
+    if (frameType === "OrdinalFrame") {
+      customProps.rExtent = this.createExtent("rExtent", state)
+    } else if (frameType === "XYFrame") {
+      customProps.xExtent = this.createExtent("xExtent", state)
+      customProps.yExtent = this.createExtent("yExtent", state)
     }
 
-    return React.cloneElement(child, { ...child.props, annotations })
+    return React.cloneElement(child, validFrameProps(customProps, frameType))
   }
 
-  render() {
-    //SHARED EXTENT
+  /**
+   * Memoize the mapping to prevent unecessary updates and not have
+   * to use the lifecycle methods.
+   */
+  processFacetController = memoize((props: Props, state: State) => {
```

Comment:
This may be a premature optimization, but instead of using `componentWillMount` and `componentWillReceiveProps`, I'm using memoization to reduce the amount of cloning, mapping, updating state done. This is recommended in React docs: https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#what-about-memoization

Reply:
Ah I hadn't seen this it'll also make it easier to get into 17.